### PR TITLE
Reverted back heuristic approach

### PR DESCRIPTION
**Pull request by Rambatino**
Shifted back to the heuristic way, optimising for low sample sizes.

### CODE REVIEW

File: CHAID/stats.py [hunk 1]
```diff
@@ -50,11 +50,15 @@ def best_split(self, ind, dep):
...
-            return self.best_cat_split(ind, dep)
+            return self.best_cat_heuristic_split(ind, dep)
 
-    def best_cat_split(self, ind, dep):
-        """ detrmine best categorical variable split """
+    def best_cat_heuristic_split(self, ind, dep):
+        """ determine best categorical variable split using heuristic methods """
         split = Split(None, None, None, None, 0)
+        if len(np.unique(dep.arr)) == 1:
+            split.invalid_reason = InvalidSplitReason.PURE_NODE
```
Comment (xulaus) [inline CHAID/stats.py]: `PURE_NODE` is a bit of a bad name. 2 hard problems I guess and it is a pre existing thing though so ... 🤷‍♂️ 

Does this conditional mean we can get rid of the previous use below?

File: CHAID/stats.py [hunk 2]
```diff
@@ -73,40 +77,103 @@ def best_cat_split(self, ind, dep):
                     for dep_v in all_dep:
                         freq[col][dep_v] = dep.weights[(ind_var.arr == col) * (dep.arr == dep_v)].sum()
 
+
+            # merge all most similar, too small categories together until have sufficient base size
+            # take the two smallest, if both smaller than min child node size, merge
+            break_out = False
+            has_merged = 1
+            while has_merged == 1:
```
Comment (xulaus) [inline CHAID/stats.py]: This kind of flagging seems super gross. Perhaps switching the last if to a negative and doing a break would be better. Also there are so many booleans flying round, some well named variables would be 👌 
```python
while True:
  data = np.array([ [k, sum(v.values())] for k,v in freq.items() ])
  data_sorted = data[data[:,1].argsort()]

  too_few_respondents =  sum(data_sorted[:, 1]) < self.min_child_node_size
  nodes_below_min = data_sorted[0:2, 1] < self.min_child_node_size
  no_splits_available = len(data_sorted[:, 1]) == 2 and nodes_below_min.any()

  if too_few_respondents or no_splits_available:
    split.invalid_reason = InvalidSplitReason.MIN_CHILD_NODE_SIZE
    break

  if not nodes_below_min.all():
    break

  first  = min(data_sorted[0, 0], data_sorted[1, 0])
  # ...
```

File: CHAID/stats.py [hunk 3]
```diff
@@ -73,40 +77,103 @@ def best_cat_split(self, ind, dep):
...
+
+                    keys = set(col1_freq.keys()).union(col2_freq.keys())
+
+                    n_ij = np.array([
+                        [col1_freq.get(k, 0) for k in keys],
+                        [col2_freq.get(k, 0) for k in keys]
+                    ])
+
+                    if n_ij.shape[1] == 1:
+                        p_split, dof, chi = 1, float('nan'), float('nan')
```
Comment (xulaus) [inline CHAID/stats.py]: `np.nan` over `float('nan')` or maybe even
```python
import np.nan as NaN
```

File: CHAID/stats.py [hunk 4]
```diff
@@ -73,40 +77,103 @@ def best_cat_split(self, ind, dep):
...
+                        break
+                    else:
+                        chi, p_split, dof = chisquare(n_ij, dep.weights is not None)
+
+                    if choice is None or p_split > highest_p_join or (p_split == highest_p_join and chi > split_chi):
+                        choice, highest_p_join, split_chi = comb, p_split, chi
+
+                invalid_reason = None
+                sufficient_split = highest_p_join < self.alpha_merge
+                if not sufficient_split: invalid_reason = InvalidSplitReason.ALPHA_MERGE
```
Comment (xulaus) [inline CHAID/stats.py]: I think dropping the `invalid_reason = None` and using `if ... elif .. else` is cleaner here

```python
if highest_p_join >= self.alpha_merge:
  split.invalid_reason = InvalidSplitReason.ALPHA_MERGE
elif any( sum(node_v.values()) < self.min_child_node_size for node_v in freq.values() ):
  split.invalid_reason = InvalidSplitReason.MIN_CHILD_NODE_SIZE
elif sufficient_split and len(freq.values()) > 1:
  # Merge loop here
```

File: tests/test_tree.py [hunk 5]
```diff
@@ -122,14 +122,34 @@ def test_best_split_with_combination():
         tree.vectorised_array,
         tree.observed
     )
-
     assert list_ordered_equal(ndarr, orig_ndarr), 'Calling chaid should have no side affects for original numpy arrays'
     assert list_ordered_equal(arr, orig_arr), 'Calling chaid should have no side affects for original numpy arrays'
     assert split.column_id == 0, 'Identifies correct column to split on'
-    assert list_unordered_equal(split.split_map, [[1], [2, 3]]), 'Correctly identifies categories'
+    assert list_unordered_equal(split.split_map, [[1], [2], [3]]), 'Correctly identifies categories'
```
Comment (xulaus) [inline tests/test_tree.py]: Are we sure?

File: tests/test_tree.py [hunk 6]
```diff
@@ -144,7 +164,7 @@ def test_new_columns_constructor():
         CHAID.OrdinalColumn(age, name="age", metadata=metadata),
     ]
     tree = CHAID.Tree(cols, CHAID.NominalColumn(income), {'min_child_node_size': 1})
-    assert tree.tree_store[0].split.groupings == "[['0-5'], ['6-10'], ['11-15']]"
+    assert tree.tree_store[0].split.groupings == "[['0-5'], ['6-10', '11-15']]"
```
Comment (xulaus) [inline tests/test_tree.py]: One again, can we properly justify this

File: CHAID/stats.py (hunk 2 above)
Comment (xulaus) [reply CHAID/stats.py]: We'll also want to hit each one of those booleans separately in its own test case.

File: CHAID/stats.py (hunk 1 above)
Comment (Rambatino) [reply CHAID/stats.py]: yes, will sort

File: tests/test_tree.py (hunk 6 above)
Comment (Rambatino) [reply tests/test_tree.py]: It's basically the difference between chi-squaring all them of them individually. As in a 3x3 chi matrix, which returns sig, or comparing each split against eachother and these two: `['6-10', '11-15']` when in a 2x2 matrix get p value > alpha merge. It's right on the level and is something that happens when your base size is like 10 and your . Rather than writing code for that specific use case, I'm happy with it merging them. The base size is 20 and won't happen if you have actual sufficient sample sizes.

File: tests/test_tree.py (hunk 5 above)
Comment (Rambatino) [reply tests/test_tree.py]: Again, same here with the small base sizes.

**Comment by Rambatino** [comment]
For the edge case:

``` python
ipdb>
([], {0: 13.0, 1: 6.0, 2: 1.0}, (age, p=0.025874595652, score=7.30898730899, groups=[['0-5'], ['6-10', '11-15']]), dof=2))
├── (['0-5'], {0: 10.0, 1: 1.0, 2: 0}, <Invalid Chaid Split> - the minimum parent node size threshold has been reached)
└── (['6-10', '11-15'], {0: 3.0, 1: 5.0, 2: 1.0}, <Invalid Chaid Split> - the minimum parent node size threshold has been reached)
```

image

**Comment by Rambatino** [comment]
Actually, we seem to be missing a tree level...

**Comment by Rambatino** [comment]
Ahh the min parent node size was wrong, here it is with 20:

image
